### PR TITLE
8309330: Allow java.security to be extended via a properties directory

### PR DESCRIPTION
**Pull request by gnu-andrew**
Currently, security properties are held within the `java.security` file in the JDK tree for each installed JDK. The system property `java.security.properties` can be used to point to a file containing additional properties. These can be appended to the existing set or override all existing properties.

There is currently no way to specify additional properties permanently or to reference multiple files. Making permanent changes to the `java.security` properties requires editing the `java.security` file in each JDK where the changes are required.

This patch allows a directory tree to be specified either permanently in the java.security file by the `security.propertiesDir` property or on the command line using `java.security.propertiesDir`. Any property files found in this directory tree can be appended to those specified in `java.security`, as with the single file used by `java.security.properties`.

As an example, the `security.propertiesDir` in the `java.security` file of each JDK can be set to a common shared directory, allowing all JDKs to share a common set of security properties. This eases setting up properties on each new JDK installation and also allows the shared properties to be maintained under different access permissions to those of the JDK.

The command-line variant, `java.security.propertiesDir`, is intended primarily for testing and to disable a permanent properties directory by setting the value to empty. As with `java.security.properties`, the system property will be ignored if `security.overridePropertiesFile` in the `java.security` file is not set to true.

A less flexible version of this patch (a permanent hardcoded single file) has been [used in our JDK installations since 2016](https://bugzilla.redhat.com/show_bug.cgi?id=1249083) to provide a system-wide crypto policy. Having support for this in the upstream JDK would allow us to remove a local patch from our builds and reduce divergence from upstream.

---------
### Progress
- [ ] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [ ] Change requires a CSR request matching fixVersion 23 to be approved (needs to be created)

### Issue
 * [JDK-8309330](https://bugs.openjdk.org/browse/JDK-8309330): Allow java.security to be extended via a properties directory (**Enhancement** - P3)

### Reviewers
 * [Alexey Bakhtin](https://openjdk.org/census#abakhtin) (@alexeybakhtin - Committer) ⚠️ Review applies to [85095b89](https://git.openjdk.org/jdk/pull/14277/files/85095b89e0f77c87e506be4e503e4d14c77e9e47)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/14277/head:pull/14277` \
`$ git checkout pull/14277`

Update a local copy of the PR: \
`$ git checkout pull/14277` \
`$ git pull https://git.openjdk.org/jdk.git pull/14277/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 14277`

View PR using the GUI difftool: \
`$ git pr show -t 14277`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/14277.diff">https://git.openjdk.org/jdk/pull/14277.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/14277#issuecomment-1572979859)

### CODE REVIEW

File: src/java.base/share/classes/java/security/Security.java [hunk 1]
```diff
@@ -104,16 +108,50 @@ private static void initialize() {
...
+        }
+
+        if (extraPropDir != null && !extraPropDir.isBlank()) {
+            if (sdebug != null ) {
+                sdebug.println("searching for property files in " + extraPropDir);
+            }
+            Stream<Path> stream = null;
+            try {
+                extraPropDir = PropertyExpander.expand(extraPropDir);
+                stream = Files.find(Path.of(extraPropDir), Integer.MAX_VALUE,
```
Comment (alexeybakhtin) [inline src/java.base/share/classes/java/security/Security.java:138]: Hi @gnu-andrew ,
The order of the files, returned by File::find is not defined. Different property files can change the same security properties, and only the latest changes will be applied. So the order of the property files is important but seems not controlled by the propertiesDir option.

File: src/java.base/share/classes/java/security/Security.java [hunk 2]
```diff
@@ -104,16 +108,50 @@ private static void initialize() {
...
+                                    ((path, attrs) -> attrs.isRegularFile()));
```
[the remark this answers is not in the document]
Comment (gnu-andrew) [reply src/java.base/share/classes/java/security/Security.java:139]: Yes, a check on `Files.isHidden(path)` should cover that.

**Review by alexeybakhtin — APPROVED** [review]
Thank you. Very useful enhancement. LGTM

**Comment by gnu-andrew** [comment]
Still waiting on CSR resolution.

**Comment by martinuy** [comment]
As commented here [1], @franferrax and I have been working on a different approach to this problem. We suggest to put this PR on hold until we have our proposal ready to compare and discuss further.

--
[1] - https://bugs.openjdk.org/browse/JDK-8309331?focusedCommentId=14606264&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-14606264

**Comment by gnu-andrew** [comment]
> As commented here [1], @franferrax and I have been working on a different approach to this problem. We suggest to put this PR on hold until we have our proposal ready to compare and discuss further.
> 
> -- [1] - https://bugs.openjdk.org/browse/JDK-8309331?focusedCommentId=14606264&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-14606264

This doesn't seem to address the same problem. This PR & CSR are aimed at extending the existing functionality of `java.security.properties` with permanent configuration and the option of multiple files, while also allowing these configuration snippets to be maintained separately outside a single JDK.  What you seem to be suggesting involves adding a meta-property to the property file to instruct it to be replaced with another file and would require multiple such instructions for each file. One of the benefits of this proposal is the directory can be setup once in the `java.security` file and then future maintenance can take place on the referenced directory for all JDKs.

**Comment by gnu-andrew** [comment]
/open
Please keep open for now.

**Pull request closed without merging**